repair/repair.cc

```diff
@@ -1713,3 +1713,48 @@ future<> decommission_with_repair(seastar::sharded<database>& db, locator::token
 future<> removenode_with_repair(seastar::sharded<database>& db, locator::token_metadata tm, gms::inet_address leaving_node) {
     return do_decommission_removenode_with_repair(db, std::move(tm), std::move(leaving_node));
 }
+
+future<> do_rebuild_replace_with_repair(seastar::sharded<database>& db, locator::token_metadata tm, sstring op, sstring source_dc) {
+    return seastar::async([&db, tm = std::move(tm), source_dc = std::move(source_dc), op = std::move(op)] () mutable {
+        auto keyspaces = db.local().get_non_system_keyspaces();
+        rlogger.info("{}: started with keyspaces={}, source_dc={}", op, keyspaces, source_dc);
+        auto myip = utils::fb_utilities::get_broadcast_address();
+        for (auto& keyspace_name : keyspaces) {
+            if (!db.local().has_keyspace(keyspace_name)) {
+                rlogger.info("{}: keyspace={} does not exist any more, ignoring it", op, keyspace_name);
+                continue;
+            }
+            auto& ks = db.local().find_keyspace(keyspace_name);
+            auto& strat = ks.get_replication_strategy();
+            dht::token_range_vector ranges = strat.get_ranges(myip);
+            std::unordered_map<dht::token_range, repair_neighbors> range_sources;
+            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}", op, keyspace_name, source_dc, ranges.size());
+            for (auto it = ranges.begin(); it != ranges.end();) {
+                auto& r = *it;
+                seastar::thread::maybe_yield();
+                auto end_token = r.end() ? r.end()->value() : dht::maximum_token();
+                auto& snitch_ptr = locator::i_endpoint_snitch::get_local_snitch_ptr();
+                auto neighbors = boost::copy_range<std::vector<gms::inet_address>>(strat.calculate_natural_endpoints(end_token, tm) |
+                    boost::adaptors::filtered([myip, &source_dc, &snitch_ptr] (const gms::inet_address& node) {
+                        if (node == myip) {
+                            return false;
+                        }
+                        return source_dc.empty() ? true : snitch_ptr->get_datacenter(node) == source_dc;
+                    })
+                );
+                rlogger.debug("{}: keyspace={}, range={}, neighbors={}", op, keyspace_name, r, neighbors);
+                if (!neighbors.empty()) {
+                    range_sources[r] = repair_neighbors(std::move(neighbors));
+                    it++;
```

Comment:
A nitpick: `++it;` would be better.

Reply:
Can you explain why

Reply:
It's because the way operator()++ is implemented in C++ compared to operator++().

Because the former has to return the value before the increment it has to read it first, increment and then return. While the later simply increments and returns the result. 

See some discussion about this here: https://softwareengineering.stackexchange.com/questions/358725/pre-increment-vs-post-increment

Reply:
Yes, I know the argument. But the iterator is for a vector and I bet the compile will optimize the return value anyway.

Reply:
Value is always returned - that has nothing to do with my comment. The problem is what needs to be done about it. In the former case you need to copy the old value aside in the later - you don't.

Reply:
My point is the compiler knows the caller of it++ or ++it does not use the value the operator returns, so it can optimize the operator function to avoid the internal copy. 
@espindola Do you have any insights?

Reply:
Depends on what 'it' is. For a simple type, sure, the result will be the same. Since it is simple to use ++it I normally find it easier to use it instead of trying to figure out if the compiler will inline and optimize.

Reply:
Thanks. Yes, makes sense to me.